File: streampipes-rest/src/main/java/org/streampipes/rest/impl/DataStream.java
/*
 * Copyright 2018 FZI Forschungszentrum Informatik
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package org.streampipes.rest.impl;

import org.streampipes.manager.operations.Operations;
import org.streampipes.model.SpDataStream;
import org.streampipes.rest.shared.serializer.annotation.GsonWithIds;

import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

@Path("/v2/users/{username}/streams")
public class DataStream extends AbstractRestInterface {

  @Path("/update")
  @POST
  @Produces(MediaType.APPLICATION_JSON)
  @GsonWithIds
  public Response getStreamsBySource(SpDataStream stream) {
    return ok(Operations.updateActualTopic(stream));
  }
}

File: streampipes-client-python/streampipes_client/endpoint/data_stream.py
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

"""
Specific implementation of the StreamPipes API's data stream endpoints.
"""
from typing import Tuple, Type

from streampipes_client.endpoint.endpoint import APIEndpoint
from streampipes_client.model.container import DataStreams

__all__ = [
    "DataStreamEndpoint",
]

from streampipes_client.model.container.resource_container import ResourceContainer


class DataStreamEndpoint(APIEndpoint):
    """Implementation of the DataStream endpoint.

    Consequently, it allows querying metadata about available data streams (see `all()` method).
    The metadata is returned as an instance of `model.container.DataStreams`.

    Parameters
    ----------
    parent_client: StreamPipesClient
        The instance of `client.StreamPipesClient` the endpoint is attached to.

    Examples
    --------

    >>> from streampipes_client.client import StreamPipesClient
    >>> from streampipes_client.client.client_config import StreamPipesClientConfig
    >>> from streampipes_client.client.credential_provider import StreamPipesApiKeyCredentials

    >>> client_config = StreamPipesClientConfig(
    ...     credential_provider=StreamPipesApiKeyCredentials(username="test-user", api_key="api-key"),
    ...     host_address="localhost",
    ...     port=8082,
    ...     https_disabled=True
    ... )

    >>> client = StreamPipesClient.create(client_config=client_config)

    >>> data_streams = client.DataStreamEndpoint.all()

    >>> len(data_streams)
    2
    """

    @property
    def _container_cls(self) -> Type[ResourceContainer]:
        """Defines the model container class the endpoint refers to.


        Returns
        -------
        `model.container.DataStreams`
        """
        return DataStreams

    @property
    def _relative_api_path(self) -> Tuple[str, ...]:
        """Defines the relative api path to the DataStream endpoint.
        Each path within the URL is defined as an own string.

        Returns
        -------
        A tuple of strings of which every represents a path value of the endpoint's API URL.
        """

        return "api", "v2", "streams"
